### PR TITLE
Add proper tests for binary search

### PR DESCRIPTION
**Pull request by CarlosZoft**
# Welcome to the JavaScript community

[Open in Gitpod](https://gitpod.io/from-referrer/)&nbsp;[know more](https://www.gitpod.io/docs/pull-requests/)

### Describe your change:

- [x] Add an algorithm?
- [x] Add some test?

### Checklist:

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Javascript/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized.
- [x] I know that pull requests will not be merged if they fail the automated tests.
- [x] This PR only changes one algorithm file. To ease review, please open separate PRs for separate algorithms.
- [x] All filenames should use the UpperCamelCase (PascalCase) style. There should be no spaces in filenames.
      **Example:**`UserProfile.js` is allowed but `userprofile.js`,`Userprofile.js`,`user-Profile.js`,`userProfile.js` are not

### CODE REVIEW

**Review by appgurueu — COMMENTED** [review]
How does this compare to the current implementation? From what I can see it's merely a less readable iterative version. If it showed better traits than the current one, you may replace it. But I don't see the merit of two largely similar iterative implementations.

**Comment by raklaptudirm** [comment]
@appgurueu since it is being added as a separate implementation, I think we could add it. It is using some more js features, which could be wanted. What do you think?

File: Search/BinarySearch.js [hunk 1]
```diff
@@ -48,8 +48,23 @@ function binarySearchIterative (arr, x, low = 0, high = arr.length - 1) {
   // if low > high => we have searched the whole array without finding the item
   return -1
 }
+function binarySearchIterativeAlternative (arr, x) {
+  let low = -1
+  let high = arr.length
+  while (low < high - 1) {
+    const half = Math.floor((low + high) / 2)
+    // arr[half] === x => low = half else high = half
```
Comment (appgurueu) [inline Search/BinarySearch.js:56]: Superfluous comment that just explains what a ternary does.

File: Search/BinarySearch.js [hunk 2]
```diff
@@ -48,8 +48,23 @@ function binarySearchIterative (arr, x, low = 0, high = arr.length - 1) {
...
+    arr[half] < x ? (low = half) : (high = half)
```
Comment (appgurueu) [inline Search/BinarySearch.js:57]: Should rather be written as `if (arr[half] < x) low = half else high = half`.

File: Search/BinarySearch.js [hunk 3]
```diff
@@ -48,8 +48,23 @@ function binarySearchIterative (arr, x, low = 0, high = arr.length - 1) {
...
+  }
+  // if arr[high] === x => return high else return -1
```
Comment (appgurueu) [inline Search/BinarySearch.js:59]: Again unneeded comment. Please don't explain what the code does, but *why* it does it (something like: "return -1 if the element doesn't match the searched element").

File: Search/BinarySearch.js [hunk 4]
```diff
@@ -89,3 +104,10 @@ export { binarySearchIterative, binarySearchRecursive }
 // binarySearchIterative(stringArr, 'Charlie')
 // binarySearchRecursive(stringArr, 'Zulu')
 // binarySearchIterative(stringArr, 'Sierra')
+// console.log("------------------------------------------")
```
Comment (appgurueu) [inline Search/BinarySearch.js:107]: This code smells like improper testing.

File: Search/BinarySearch.js [hunk 5]
```diff
@@ -48,8 +48,23 @@ function binarySearchIterative (arr, x, low = 0, high = arr.length - 1) {
   // if low > high => we have searched the whole array without finding the item
   return -1
 }
+function binarySearchIterativeAlternative (arr, x) {
+  let low = -1
+  let high = arr.length
+  while (low < high - 1) {
+    const half = Math.floor((low + high) / 2)
```
Comment (appgurueu) [inline Search/BinarySearch.js:55]: This is a minor simplification over what the other implementation does. Consider opening a separate PR, changing the other implementation to `const mid = Math.floor((low + high) / 2)`.

File: Search/BinarySearch.js [hunk 6]
```diff
@@ -48,8 +48,23 @@ function binarySearchIterative (arr, x, low = 0, high = arr.length - 1) {
   // if low > high => we have searched the whole array without finding the item
   return -1
 }
+function binarySearchIterativeAlternative (arr, x) {
+  let low = -1
```
Comment (appgurueu) [inline Search/BinarySearch.js:52]: Consider adding a comment stating that these bounds are both exclusive (which is probably less intuitive than the other impl.).

**Review by appgurueu — CHANGES_REQUESTED** [review]
This PR may indeed be using more JS features (primarily ternaries) but in a wrong manner; they only make the code less readable (except probably for the last ternary). Currently I see only two very tiny advantages over the other impl., the simplification of the term for `mid` and moving the check if found out of the loop; I don't see much of a benefit from the exclusive low & high indices, which is a mere convention. On the other hand however there is rather unreadable code, bad comments and syntactic sugar where it doesn't belong.

The tiny improvements should just be changes to the current implementation.

File: Search/BinarySearch.js [hunk 7]
```diff
@@ -48,8 +48,25 @@ function binarySearchIterative (arr, x, low = 0, high = arr.length - 1) {
...
+}
 
-export { binarySearchIterative, binarySearchRecursive }
+export {
+  binarySearchIterative,
+  binarySearchIterativeAlternative,
+  binarySearchRecursive
+}
 
 /* ---------------------------------- Test ---------------------------------- */
```
Comment (raklaptudirm) [inline Search/BinarySearch.js:71]: Remove these comments and replace them with proper jest tests.

**Review by raklaptudirm — CHANGES_REQUESTED** [review]

**Comment by appgurueu** [comment]
At this point it's similar enough that you should just tidy up the existing implementation (adding proper jest tests) instead.

**Comment by CarlosZoft** [comment]
I did this

**Review by appgurueu — CHANGES_REQUESTED** [review]
From a quick glance the tests look fine to me. Definitely a significant improvement. But please deduplicate the code; you can do something like `funcs = {iterative: binarySearchIterative, recursive: binarySearchRecursive}; for (let name in funcs) runTests(name, funcs[name])` where `runTests` is a parameterized function that runs the tests (which are the same for both implementations).

**Comment by CarlosZoft** [comment]
Nice vision, i will make this!

**Comment by CarlosZoft** [comment]
lgtm now , what you think?

**Review by appgurueu — APPROVED** [review]
Thank you for your contribution. LGTM!

**Review by raklaptudirm — APPROVED** [review]

**Pull request merged**